pkg/config/xds/apiclient/grpc.go

```
@@ -261,3 +277,103 @@ func (g *GrpcApiClient) init() {
+	clusters  *sync.Map // map[clusterName]*grpcCluster
+	clusterMg controls.ClusterManager
+}
+
+type GrpcCluster struct {
+	name   string //cluster name
+	config *model.Cluster
+	once   sync.Once
+	conn   *grpc.ClientConn
+}
+
+// GetGrpcCluster get the cluster or create it first time.
+func (g *GrpcClusterManager) GetGrpcCluster(name string) (*GrpcCluster, error) {
+	if !g.clusterMg.HasCluster(name) {
+		return nil, errors.Wrapf(ErrClusterNotFound, "name = %s", name)
+	}
+
+	if load, ok := g.clusters.Load(name); ok {
+		grpcCluster := load.(*GrpcCluster) // grpcClusterManager only
+		return grpcCluster, nil
+	}
+
+	store, err := g.clusterMg.CloneXdsControlStore()
+	if err != nil {
+		return nil, errors.WithMessagef(err, "clone cluster store failed")
+	}
+
+	var clusterCfg *model.Cluster
+	for _, cfg := range store.Config() {
+		if cfg.Name == name {
+			clusterCfg = cfg
+			break
+		}
+	}
+	if clusterCfg == nil {
+		return nil, errors.Wrapf(ErrClusterNotFound, "name of %s", name)
+	}
+	newCluster := &GrpcCluster{
+		name:   name,
+		config: clusterCfg,
+	}
+	g.clusters.Store(name, newCluster)
+	return newCluster, nil
+}
+
+func (g *GrpcClusterManager) Close() (err error) {
+	//todo enhance the close process when concurrent
+	g.clusters.Range(func(_, value interface{}) bool {
+		if conn := value.(*grpc.ClientConn); conn != nil {
+			if err = conn.Close(); err != nil {
+				logger.Errorf("can not close grpc connection.", err)
+			}
+		}
+		return true
+	})
+	return nil
+}
+
+func (g *GrpcCluster) GetConnect() *grpc.ClientConn {
+	g.once.Do(func() {
+		creds := insecure.NewCredentials()
+		//if *xdsCreds { // todo
+		//	log.Println("Using xDS credentials...")
+		//	var err error
+		//	if creds, err = xdscreds.NewClientCredentials(xdscreds.ClientOptions{FallbackCreds: insecure.NewCredentials()}); err != nil {
+		//		log.Fatalf("failed to create client-side xDS credentials: %v", err)
+		//	}
+		//}
+		if len(g.config.Endpoints) == 0 {
+			panic("expect endpoint.")
```

这里为啥要 panic？在 once 函数里 panic 合适吗？你这个函数都不确定在那个 goroutine 里面运行，一旦把整个程序搞崩合适吗？

---

config 缺少对应的配置，导致整个grpc 无法建立连接，动态配置无法按照配置启动，所以整个系统应该fail early